### PR TITLE
feat(FEC-10973): parse image tracks from dash manifest

### PR DESCRIPTION
**Pull request by dan-ziv**
### Description of the Changes

Shaka player does not support parsing of image type adaptation set, so need to create our own dedicated parser for this.
We will register to the response filter and extract the image tracks data in the following raw structure:

- AdaptationSet
    - ?EssentialProperty
    - SegmentTemplate
    - List<Representation>
      - ?EssentialProperty

See diagram for better understanding of the flow:

DashManifestParser flow (1)

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

---
This change is [Reviewable](https://reviewable.io/reviews/kaltura/playkit-js-dash/131)

### CODE REVIEW

File: src/parser/parser-utils.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    const domParser = new DOMParser();
+    return domParser.parseFromString(text, 'text/xml');
+  },
+  parseAttr: (elem: HTMLElement, name: string, parseFunction?: Function, defaultValue: any): any => {
+    let parsedValue = null;
+    const value = elem.getAttribute(name);
+    if (value !== null) {
+      parsedValue = parseFunction ? parseFunction(value) : value;
+    }
+    return parsedValue == null ? defaultValue : parsedValue;
```
Comment (yairans) [inline src/parser/parser-utils.js:49]: dont see `defaultValue` usage is it for the future?

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/parser/parser-utils.js (hunk 1 above)
Comment (dan-ziv) [reply src/parser/parser-utils.js:49]: yes

File: src/parser/parser-utils.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  },
+  findElement: (node: HTMLElement, name: string): HTMLCollection<HTMLElement> => {
```
[the remark this answers is not in the document]
Comment (dan-ziv) [reply src/parser/parser-utils.js:51]: fixed

File: src/parser/dash-manifest-parser.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    } catch (e) {
+      this._logger.error('Manifest parsing failed', e);
+    }
+  }
+
+  getAdaptationSet(type: string): ?AdaptationSet {
+    if (type) {
+      const adaptationSets = this._adaptationSets.filter((adaptationSet: AdaptationSet) => adaptationSet.contentType === type);
+      if (adaptationSets) {
+        return adaptationSets[0];
```
[the remark this answers is not in the document]
Comment (dan-ziv) [reply src/parser/dash-manifest-parser.js:33]: fixed

[the remark this answers is not in the document]
Comment (dan-ziv) [reply src/parser/dash-manifest-parser.js:33]: fixed

File: src/parser/dash-manifest-parser.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  parseManifest() {
+    try {
+      this._logger.debug('Start parsing dash manifest');
+      // For now parse only adaptation sets
+      this._parseAdaptionSets();
+      this._logger.debug('Manifest parsing finished successfully');
+    } catch (e) {
+      this._logger.error('Manifest parsing failed', e);
```
[the remark this answers is not in the document]
Comment (dan-ziv) [reply src/parser/dash-manifest-parser.js:25]: ok

File: src/parser/parser-utils.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    return node.getElementsByTagName(name);
+  },
+  findChild(elem: HTMLElement, name: string): ?HTMLElement {
+    const children = this.findChildren(elem, name);
+    if (children.length !== 1) {
+      return null;
```
[the remark this answers is not in the document]
Comment (dan-ziv) [reply src/parser/parser-utils.js:57]: used when only one child should be (like essential property, segment template, etc), if more than one returned this is not valid.

File: src/parser/parser-utils.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
+// @flow
+const ParserUtils = {
+  BufferToStr: (buffer: ArrayBuffer): string => {
+    const textDecoder = new TextDecoder();
+    return textDecoder.decode(buffer);
+  },
+  StrToBuffer: (str: string): Uint8Array => {
+    const textEncoder = new TextEncoder();
+    return textEncoder.encode(str);
+  }
```
Comment (OrenMe) [inline src/parser/parser-utils.js:10]: TextDecoder and TextDecoder are not supported in IE11.
Either we say we don't support it on IE11 and add protections by just checking if TextDecoder and TextEncoder are in window.
Another option is to add a polyfill: https://gist.github.com/Yaffle/5458286

**Review by OrenMe — CHANGES_REQUESTED** [review]

File: src/parser/parser-utils.js (hunk 6 above)
Comment (dan-ziv) [reply src/parser/parser-utils.js]: fixed

**Review by OrenMe — DISMISSED** [review]

File: src/dash-adapter.js [hunk 7]
```diff
@@ -620,12 +628,26 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
           });
           break;
         case shaka.net.NetworkingEngine.RequestType.MANIFEST:
+          this._parseManifest(response.data);
           this._trigger(EventType.MANIFEST_LOADED, {miliSeconds: response.timeMs});
-          break;
```
Comment (yairans) [inline src/dash-adapter.js:624]: I would keep it to be on the safe side.
if tomorrow someone will add new case it will be a bug

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/dash-adapter.js (hunk 7 above)
Comment (dan-ziv) [reply src/dash-adapter.js:624]: probably removed by mistake.

**Review by yairans — APPROVED** [review]

**Review by OrenMe — APPROVED** [review]

**Pull request merged**